add "/apis/" to kube-aggregator apisHandler

CaoShuFeng opened the pull request:
This makes the following two urls have the same result.
https://ip:443/apis
https://ip:443/apis/
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44021)
<!-- Reviewable:end -->

CaoShuFeng commented:
@deads2k

liggitt commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -178,6 +178,7 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 		endpointsLister: s.endpointsLister,
 	}
 	s.GenericAPIServer.HandlerContainer.Handle("/apis", apisHandler)
+	s.GenericAPIServer.HandlerContainer.Handle("/apis/", apisHandler)
```
this has the side-effect of making any subpath of `/apis/` (like `/apis/foo/bar/baz`) be handled by `apisHandler`. Is that what we want?

CaoShuFeng replied:
Thanks for your review.
I test this pr with both legal and illegal URLs with my patch.
I can't find any difference except when  accessing "/apis/". I thinks this is because the subpath has been registered in other places.
Without this pr when accessing "/apis" we use this handler:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go#L77
but when accessing "/apis/"  we use this handler:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/discovery.go#L35
And these tow handlers return different values in my cluster.

CaoShuFeng replied:
This provides consistency with line 224, I think.

deads2k replied:
> This provides consistency with line 224, I think.

That line does something different.  It's more narrowly scoped.



> this has the side-effect of making any subpath of /apis/ (like /apis/foo/bar/baz) be handled by apisHandler. Is that what we want?

There's a double check to make sure that we delegate through if there's a subpath.

liggitt replied:
Even subpaths that are not handled by specifically registered handlers?

CaoShuFeng replied:
> Even subpaths that are not handled by specifically registered handlers?

We will use FallThroughHandler for URLs that are not /apis or /apis/
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go#L83

CaoShuFeng committed:
add "/apis/" to kube-aggregator apisHandler

This makes the following two urls have the same result.
https://<ip>:443/apis
https://<ip>:443/apis/

deads2k commented:
/lgtm

deads2k commented:
@k8s-bot ok to test

CaoShuFeng commented:
@k8s-bot verify test this
